# Data and Code SOPs

## General Procedures

* [Coding Best Practices](Best_Practices_Coding)
* Electronic Lab Noteboooks
* [How to Create an SOP](How_to_create_an_SOP)

## Data deposition

* [Depositing Data in NCBI](NCBI_deposit) - including BioProject, BioSample, Genome, and SRA deposition
* [Data retrieval from public databases](Data_retrieval)

## Genome Analysis
Before annotating or assembling it is a good idea to start the BioSample and BioProject (once you know the sample is clean)

* [Fungal Genome_Assembly_with_Illumina_Reads](Genome_Assembly_with_Illumina_Reads)
* [Fungal Genome annotation](Fungal_Genome_annotation)
* [Hybrid Assembly](Hybrid_genome_assembly)
* [Contamination assessment](Blobtools)

## Amplicon Analysis

* [UCR HPCC AMPtk amplicon (microbiome) analysis](AMPtk)
* 16S Bacteria and Archaea
* ITS Fungal

## Metagenome assembly and analysis

* [KBase Metagenomes Analysis](KBase)
* UCR HPCC Metagenomes Analysis
  * Data cleaning
  * Metagenome Assembly
  * Metagenome annotation and Analysis

## Viral metagnomics
* [KBase Viral Analysis](KBase_Viruses)
* [UCR HPCC Cluster Viral Analysis](Cluster_Viruses)

## Phylogenomics

* [PHYling V1 phylogenomics](Phylogenomics_with_PHYling_V1)
* [PHYling V2 phylogenomics](https://github.com/stajichlab/PHYling)

## Transcriptome

* [Trinity RNA-seq Assembly](Trinity_RNASeq)

## Vizualization

* [JBrowse](Jbrowse_Instructions)
