# Using KBase for viral discovery and annotation from metagenomes

KBase is an online web tool from the Department of Energy that is focused on providing bioinformatics infrastructure for genomics and metagenomics related anlayses. We have found that this tool works well for undergraduate researchers and those new to bioinformatic analyses as a stepping stone to command line work.
KBase can be accessed by creating an account or signing in through one's ORCID. Within KBase the [Stajich Lab] (https://narrative.kbase.us/#/orgs/stajichlab) has a group that one can share and associate narratives with. Narratives are essentially IPython notebooks (or bioinformatic labratory notebooks) which contain all of the KBase modules that are run and what data is used as well as any resulting outputs. KBase narratives can also be 'forked' or copied and then edited allowing for users to re-use or edit a narrative as a jumping off point for their own analyses.

The viral protocol that we follow is currently provided as a [KBase tutorial narrative 'Viral End-To-End Analysis'](https://kbase.us/n/75811/85/) which is the current protcol we use in lab.
An almost identical protcol is on [Protocols.io](https://www.protocols.io/view/processing-a-viral-metagenome-using-ivirus-buisnuee), but which uses CyVerse which our undergraduates found harder to work with.
Dr. Benjamin Bolduc also gives a webinar on use of this narrative through KBase which is on [YouTube](https://www.youtube.com/watch?v=WbQeOzdyTbc) and which is highly recommeded for understanding the innerworking of the KBase modules used and also how to interpret the VirSorter and VContact2 results.

Currently using the VContactv2 beta application will enable you to download the resulting network for visualization in Cytoscape (see Protocols.io tutorials for Cytoscape instructions). 
